Sources/SwiftSyntax/SyntaxVisitor.swift.gyb

```diff
@@ -141,3 +141,75 @@ open class SyntaxVisitor {
     }
   }
 }
+
+public protocol SyntaxTransformVisitor {
+  associatedtype ResultType = Void
+  
+  func visit(_ token: TokenSyntax) -> [ResultType]
+  func visit(_ node: UnknownSyntax) -> [ResultType]
+  
+% for node in SYNTAX_NODES:
+%   if is_visitable(node):
+  /// Visiting `${node.name}` specifically.
+  ///   - Parameter node: the node we are visiting.
+  ///   - Returns: the sum of whatever the child visitors return.
+  func visit(_ node: ${node.name}) -> [ResultType]
+%   end
+% end
+}
+
+extension SyntaxTransformVisitor {
+  public func visit(_ token: TokenSyntax) -> [ResultType] { [] }
+  public func visit(_ node: UnknownSyntax) -> [ResultType] { [] }
+
+% for node in SYNTAX_NODES:
+%   if is_visitable(node):
+  /// Visiting `${node.name}` specifically.
+  ///   - Parameter node: the node we are visiting.
+  ///   - Returns: nil by default.
+  public func visit(_ node: ${node.name}) -> [ResultType] {
+%   if node.is_base():
+      // Avoid calling into visitChildren if possible.
+      if !node.raw.layoutView!.children.isEmpty {
+        return visitChildren(node)
+      }
+      return []
+%   else:
+      // Avoid calling into visitChildren if possible.
+      if !node.raw.layoutView!.children.isEmpty {
+        return visitChildren(node)
+      }
+      return []
+%   end
+  }
+%   end
+% end
+
+  public func visit(_ data: SyntaxData) -> [ResultType] {
+    switch data.raw.kind {
+    case .token:
+      let node = TokenSyntax(data)
+      return visit(node)
+    case .unknown:
+      let node = UnknownSyntax(data)
+      return visit(node)
+    // The implementation of every generated case goes into its own function. This
+    // circumvents an issue where the compiler allocates stack space for every
+    // case statement next to each other in debug builds, causing it to allocate
+    // ~50KB per call to this function. rdar://55929175
+  % for node in NON_BASE_SYNTAX_NODES:
+    case .${node.swift_syntax_kind}:
+      let node = ${node.name}(data)
+      return visit(node)
+  % end
+    }
+  }
+
+  public func visitChildren<SyntaxType: SyntaxProtocol>(_ node: SyntaxType) -> [ResultType] {
```

Comment:
Similarly to `visit(_ data: SyntaxData)`, this should be `private`.

Reply:
Why? It is useful to API consumers (see the FuncCounter test).